client/session/fssession/send.go

```diff
@@ -0,0 +1,211 @@
+	if sp.doneCh != nil {
+		doneCh = sp.doneCh
+		sp.doneCh = nil
+	}
+	err := pr.sendFn(stream, sp.root, excludes, progress)
+	if doneCh != nil {
+		if err != nil {
+			doneCh <- err
+		}
+		close(doneCh)
+	}
+	return err
+}
+
+func (sp *fsSendProvider) SetNextProgressCallback(f func(int, bool), doneCh chan error) {
+	sp.p = f
+	sp.doneCh = doneCh
+}
+
+type progressCb func(int, bool)
+
+type protocol struct {
+	name   string
+	sendFn func(stream grpc.Stream, srcDir string, excludes []string, progress progressCb) error
+	recvFn func(stream grpc.Stream, destDir string, cu CacheUpdater) error
+}
+
+func isProtoSupported(p string) bool {
+	if override := os.Getenv("BUILD_STREAM_PROTOCOL"); override != "" {
+		return strings.EqualFold(p, override)
+	}
+	return true
+}
+
+var supportedProtocols = []protocol{
+	{
+		name:   "tarstream",
+		sendFn: sendTarStream,
+		recvFn: recvTarStream,
+	},
+	{
+		name:   "diffcopy",
+		sendFn: sendDiffCopy,
+		recvFn: recvDiffCopy,
+	},
+}
+
+// FSSendRequestOpt defines options for FSSend request
+type FSSendRequestOpt struct {
+	SrcPaths         []string
+	OverrideExcludes bool
+	DestDir          string
+	CacheUpdater     CacheUpdater
+}
+
+// CacheUpdater is an object capable of sending notifications for the cache hash changes
+type CacheUpdater interface {
+	MarkSupported(bool)
+	HandleChange(fsutil.ChangeKind, string, os.FileInfo, error) error
+}
+
+// FSSend initializes a transfer of files
+func FSSend(ctx context.Context, name string, c session.Caller, opt FSSendRequestOpt) error {
+	var pr *protocol
+	for _, p := range supportedProtocols {
+		if isProtoSupported(p.name) && c.Supports(session.MethodURL(_FileSync_serviceDesc.ServiceName+"."+name, p.name)) {
+			pr = &p
+			break
+		}
+	}
+	if pr == nil {
+		return errors.Errorf("no fssend handlers for %s", name)
+	}
+
+	opts := make(map[string][]string)
+	if opt.OverrideExcludes {
+		opts["Override-Excludes"] = []string{"true"}
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	client := newFileSyncClientWithName(name, c.Conn())
+
+	var stream grpc.ClientStream
+
+	ctx = metadata.NewContext(ctx, opts)
+
+	switch pr.name {
+	case "tarstream":
+		cc, err := client.TarStream(ctx)
+		if err != nil {
+			return err
+		}
+		stream = cc
+	case "diffcopy":
+		cc, err := client.DiffCopy(ctx)
+		if err != nil {
+			return err
+		}
+		stream = cc
+	}
+
+	return pr.recvFn(stream, opt.DestDir, opt.CacheUpdater)
+}
+
+// TODO: generate this, or use clientinterceptor
+type namedFileSyncClient struct {
+	cc   *grpc.ClientConn
+	name string
+}
+
+func newFileSyncClientWithName(name string, cc *grpc.ClientConn) FileSyncClient {
+	return &namedFileSyncClient{cc, name}
+}
+
+func (c *namedFileSyncClient) DiffCopy(ctx context.Context, opts ...grpc.CallOption) (FileSync_DiffCopyClient, error) {
+	stream, err := grpc.NewClientStream(ctx, &_FileSync_serviceDesc.Streams[0], c.cc, "/"+_FileSync_serviceDesc.ServiceName+"."+c.name+"/DiffCopy", opts...)
```

Comment:
It really doesn't make us incompatible with anything. The implementation was switched to grpc for exactly the reason that anyone could make their own implementation, in any language or using any method. And later it was switched to server handlers so that separate proto could be defined for every method. gRPC is a specification for using http2 requests for rpc and `go-grpc` is a go api for implementing clients and handlers. We use both exactly like it is recommended. A specific generated code from a specific tool is not a standard.

Concerning maintenance, this will likely remain the only place where this is ever needed. All the other usages of the session that I'm aware of don't require it. We could generate it but it really is only these lines. Only maintenance issue would be if `grpc-go` package changes its function signatures after revendoring and it makes this code not compile same way as it makes all the generated code not compile. I could also solve it with an interceptor, wrapping one client into another without any duplicate grpc pkg calls but I think it would be a lot less readable and leak to everything that uses this connection. 

Anyone reimplementing their own client/handler for even this specific feature do not need to do the similar thing. You can just define the full name in proto. We need it for a specific case because we don't know user's configuration ahead of time and can't generate proto dynamically.

There is another way to solve it(without going back to the generic approach). It means sending the name part in the request. The downside is that it makes detecting supported features very hacky. While atm the grpc server directly exposes this info we would now need to put this together manually, injecting the names so they can be detected by the remote side. It means hardcoding the "exported names" concept to the hijack layer, while atm it doesn't know anything about the requests or transport.